### PR TITLE
feat: set unique meta titles

### LINKED ISSUE #908: Set unique meta title tags
**Summary**

Per https://moz.com/learn/seo/title-tag, every page should have a unique page titles for SEO. The most important part of the title should be first, as that's shown in browser tabs. They should be 50 - 60 characters in length, but that's not as important; we can go over 60 characters.

This is great:

When we have page tabs, the meta title should be `Color - Accessibility - Carbon Design System`:

Let's include the library in the asset meta titles, so this example should be `Alluvial - Carbon Charts Svelte - Carbon Design System`:

Also `Library assets - Carbon Charts Svelte - Carbon Design System`, for example:

We don't need to include library versions in meta titles because in a separate issue, we'll set canonical links so only the "latest" version of library and asset web pages will get indexed by search engines.

Also, this information could be useful for our search service for when we provide a site search.

**Acceptance criteria**

What requirements must be met to complete this request?

1. All pages have unique page titles

### LINKED ISSUE #1576: Asset page tabs are defaulting to latest ref
**Description**

> Tell us more about the problem that you're running into. What did you see? What did you expect to
> see?

The page tabs for asset docs should preserve the ref.

**Steps to reproduce**

> How do we reproduce the error? Please be as specific as you can.

1. Go here: https://next.carbondesignsystem.com/libraries/carbon-react/v11.21.0/assets/accordion
2. Click the Usage page tab
3. That takes you to: https://next.carbondesignsystem.com/libraries/carbon-react/latest/assets/accordion/usage

The ref should be preserved, so the URL should be:

https://next.carbondesignsystem.com/libraries/carbon-react/v11.21.0/assets/accordion/usage

### PR DESCRIPTION
**Pull request by mattrosno**
Closes #908 
Closes #1576 

The PR makes sure all pages in the web app have unique meta titles. This is good for SEO, but more important to us right now, provides better reporting in Google Analytics.

#### Changelog

**New**

- Accidentally committed a `/public/data` JSON cache file. It's only open source data. Not a big deal. At some point we should flush this development data and repopulate it with newer data to cache.
- `/utils/seo.js` with `getMetaTitle` to construct meta titles based on nav data
- `/utils/use-meta-title.js` with `useMetaTitle` hook to get path-specific meta titles referencing nav data

**Changed**

- The `WithLoading` component now says `Loading...` in the meta title while the page loads
- The web app has a more complete meta description
- `/services/web-app/pages/libraries/[host]/[org]/[repo]/[library]/[ref]/index.js` was pretty messy, so I grouped some of the constants and hooks and render functions and added comments so that file is more readable

**Removed**

- N/A

#### Testing / reviewing

All page header titles should remain unchanged in the PR.

You might need to delete `/public/data` JSON files when testing library versions.

Hit every page. The meta titles should be driven by the nav data, with the most specific nav data node being first in the `-` delineated meta title.

Test MDX pages that have tabs. The tab name should come first in the meta title.

Any page in a library (/assets, /design-kits, an asset, asset docs) should include the library name in the meta title. If viewing a specific library with any of those pages, the library version is in the meta title.

### CODE REVIEW

File: services/web-app/utils/seo.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+              title = `${data3.title} - ${data2.title} - ${data1.title}`
+            }
+          })
+        }
+      })
+    }
+  })
+
+  return title
+}
```
Comment (francinelucca) [inline services/web-app/utils/seo.js:66]: To call the recursive function:
```suggestion
/**
 * For a given path like `/about-carbon/how-carbon-works`, if that matches any known nav data,
 * construct a title like "How Carbon works - About Carbon" to ensure all titles are unique. The
 * lowest node in the tree is shown first in the meta title.
 *
 * @param {string} path - typically next/router `asPath`
 * @param {Array} navData - optional nav data for dynamic side navs, e.g. libraries
 * @param {boolean} hasTabs - if the page has page tabs
 * @returns {string}
 */
export const getMetaTitle = (path, navData = [], hasTabs = false) => {
  const allNavData = [...globalNavData, ...standardsNavData, ...assetsNavData, ...navData]

  // if page tabs, pop the last path for the path to match, and all the paths in nav data
  const sideNavPath = getSideNavPath(path, hasTabs)

  return getPageTitle(allNavData, sideNavPath, hasTabs)
}
```

File: services/web-app/utils/seo.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ * @returns {string}
+ */
+const getPath = (path, popPath) => {
+  if (popPath) {
+    return path?.split('/')?.slice(0, -1)?.join('/')
+  }
+
+  return path
+}
+
```
Comment (francinelucca) [inline services/web-app/utils/seo.js:24]: I feel like doing it recursively is a bit easier to read in this case?, totally optional!:
```suggestion

/**
 * For a given side nav path like `/about-carbon/how-carbon-works`,
 *  if that matches any supplied nav data,
 * construct a title like "How Carbon works - About Carbon" to ensure all titles are unique. The
 * lowest node in the tree is shown first in the meta title.
 *
 * @param {Array} navData - array of all nav items to find sideNavPath in
 * @param {string} sideNavPath - path to side nav item to find
 * @param {boolean} hasTabs - if the page has page tabs
 * @returns {string} title
 */
const getPageTitle = (navData, sideNavPath, hasTabs, parentTitle = '') => {
  for (const navItem of navData) {
    const navItemTitle = parentTitle ? `${navItem.title} - ${parentTitle}` : navItem.title
    if (getSideNavPath(navItem.path, hasTabs) === sideNavPath) {
      return navItemTitle
    }
    if (Array.isArray(navItem.items)) {
      const title = getPageTitle(navItem.items, sideNavPath, hasTabs, navItemTitle)
      if (title) {
        return title
      }
    }
  }
}

```

File: services/web-app/components/mdx-page/mdx-page.js [hunk 3]
```diff
@@ -92,14 +80,28 @@ const MdxPage = ({
...
+  const tabsData = tabs.map((tab) => {
+    if (tab?.name && tab?.path) {
+      return tab
+    }
+    const tabSlug = slugify(tab, { strict: true, lower: true })
+    return {
+      name: tab,
+      path: baseSegment ? `/${baseSegment}/${tabSlug}` : `/${tabSlug}`
+    }
+  })
```
Comment (francinelucca) [inline services/web-app/components/mdx-page/mdx-page.js:103]: Tabs can be not supplied: https://github.com/carbon-design-system/carbon-platform/blob/main/services/web-app/pages/libraries/%5Bhost%5D/%5Borg%5D/%5Brepo%5D/%5Blibrary%5D/%5Bref%5D/assets/%5Basset%5D/%5Btab%5D.js#L86
```suggestion
  const tabsData = tabs?.map((tab) => {
    if (tab?.name && tab?.path) {
      return tab
    }
    const tabSlug = slugify(tab, { strict: true, lower: true })
    return {
      name: tab,
      path: baseSegment ? `/${baseSegment}/${tabSlug}` : `/${tabSlug}`
    }
  })
```

**Review by francinelucca — CHANGES_REQUESTED** [review]
LGTM!, one minor observation an some cleanup suggestions 🚀

**Comment by mattrosno** [comment]
@carbon-design-system/carbon-platform-devs this seems to be working great with @francinelucca's changes. I think I hit every page and the meta titles are working. If you want to test all pages in a library at a version to see that in the meta title:

http://localhost:3000/libraries/carbon-react/v11.21.0

**Review by francinelucca — APPROVED** [review]
🚀

**Review by alisonjoseph — APPROVED** [review]
LGTM!

**Pull request merged**